plugins/common/adx/adx.go

```diff
@@ -0,0 +1,206 @@
+package adx
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-kusto-go/kusto"
+	kustoerrors "github.com/Azure/azure-kusto-go/kusto/data/errors"
+	"github.com/Azure/azure-kusto-go/kusto/ingest"
+	"github.com/Azure/azure-kusto-go/kusto/kql"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/internal/choice"
+)
+
+type Config struct {
+	Endpoint        string          `toml:"endpoint_url"`
+	Database        string          `toml:"database"`
+	Timeout         config.Duration `toml:"timeout"`
+	MetricsGrouping string          `toml:"metrics_grouping_type"`
+	TableName       string          `toml:"table_name"`
+	CreateTables    bool            `toml:"create_tables"`
+	IngestionType   string          `toml:"ingestion_type"`
+}
+
+type Client struct {
+	cfg       *Config
+	conn      *kusto.ConnectionStringBuilder
+	client    *kusto.Client
+	ingestors map[string]ingest.Ingestor
+	logger    telegraf.Logger
+}
+
+func (cfg *Config) NewClient(app string, log telegraf.Logger) (*Client, error) {
+	if cfg.Endpoint == "" {
+		return nil, errors.New("endpoint configuration cannot be empty")
+	}
+	if cfg.Database == "" {
+		return nil, errors.New("database configuration cannot be empty")
+	}
+
+	cfg.MetricsGrouping = strings.ToLower(cfg.MetricsGrouping)
+	if cfg.MetricsGrouping == SingleTable && cfg.TableName == "" {
+		return nil, errors.New("table name cannot be empty for SingleTable metrics grouping type")
+	}
+
+	if cfg.MetricsGrouping == "" {
+		cfg.MetricsGrouping = TablePerMetric
+	}
+
+	if !(cfg.MetricsGrouping == SingleTable || cfg.MetricsGrouping == TablePerMetric) {
+		return nil, errors.New("metrics grouping type is not valid")
+	}
+
+	if cfg.Timeout == 0 {
+		cfg.Timeout = config.Duration(20 * time.Second)
+	}
+
+	if cfg.IngestionType == "" {
+		cfg.IngestionType = QueuedIngestion
+	} else if !(choice.Contains(cfg.IngestionType, []string{ManagedIngestion, QueuedIngestion})) {
+		return nil, fmt.Errorf("unknown ingestion type %q", cfg.IngestionType)
+	}
```

Comment:
This might be easier to read
```suggestion
	switch cfg.IngestionType {
	case "":
		cfg.IngestionType = QueuedIngestion
	case ManagedIngestion, QueuedIngestion:
		// Do nothing as those are valid
	default:
		return nil, fmt.Errorf("unknown ingestion type %q", cfg.IngestionType)
	}
```